source/material_model/compositing.cc

```diff
@@ -129,44 +123,37 @@ namespace aspect
               AssertThrow(model_name != "compositing",
                           ExcMessage("You may not use ``compositing'' as the base model for the "
                                      + prop_it->first +" property of a compositing material model."));
-              unsigned int model_ind;
-              for (model_ind=0; model_ind < model_names.size(); ++model_ind)
+              std::vector<std::string>::iterator model_position = std::find(model_names.begin(), model_names.end(), model_name);
+              if ( model_position == model_names.end() )
                 {
-                  if (model_names[model_ind] == model_name)
-                    break;
+                  model_property_map[prop] = model_names.size();
+                  model_names.push_back(model_name);
                 }
-              if (model_ind == model_names.size())
-                model_names.push_back(model_name);
-
-              model_property_map[prop] = model_ind;
+              else
+                model_property_map[prop] = std::distance(model_names.begin(), model_position);
             }
           models.resize(model_names.size());
 
-          // create the model and initialize their SimulatorAccess base
-          // class; it will get a chance to read its parameters below after we
-          // leave the current section
-          for (unsigned int i=0; i<model_names.size(); ++i)
-            {
-              models.at(i).reset(create_material_model<dim>(model_names[i]));
-              if (SimulatorAccess<dim> *sim = dynamic_cast<SimulatorAccess<dim>*>(models.at(i).get()))
-                sim->initialize_simulator (this->get_simulator());
-            }
         }
         prm.leave_subsection();
       }
       prm.leave_subsection();
 
-      /* After parsing the parameters for averaging, it is essential to parse
-      parameters related to the base model. */
+      // create the models and initialize their SimulatorAccess base
+      // After parsing the parameters for averaging, it is essential to parse
+      // parameters related to the base models
       for (unsigned int i=0; i<model_names.size(); ++i)
         {
+          models.at(i).reset(create_material_model<dim>(model_names[i]));
```

Comment:
One more nitpick (I just mention it because I have the above comment anyway): Now you resize `models` above (probably for speed?), but then use `at` here instead of `[]` further down, probably just copied from the earlier loop? Two comments about that:
1. Please move the resize command in front of this loop, where `models` is used, so that related commands are close together.
2. Use consistently `at` or `[]` do not mix between them. In this case `[]` is totally safe, because `i` will never exceed the bounds of the vector. In general `at` is significantly slower (see e.g. https://stackoverflow.com/questions/9376049/vectorat-vs-vectoroperator), although of course in this case with just a few entries this does not matter at all.